src/nvim/terminal.c

```diff
@@ -359,11 +359,13 @@ void terminal_enter(void)
   // Disable these options in terminal-mode. They are nonsense because cursor is
   // placed at end of buffer to "follow" output. #11072
   win_T *save_curwin = curwin;
-  int save_w_p_cul = curwin->w_p_cul;
+  char_u *save_w_p_culopt = curwin->w_p_culopt;
+  char_u save_w_p_culopt_flags = curwin->w_p_culopt_flags;
   int save_w_p_cuc = curwin->w_p_cuc;
   long save_w_p_so = curwin->w_p_so;
   long save_w_p_siso = curwin->w_p_siso;
-  curwin->w_p_cul = false;
+  curwin->w_p_culopt = (char_u *)xstrdup("number");
+  curwin->w_p_culopt_flags = CULOPT_NBR;
```

Comment:
Let's check the current value of `curwin->w_p_culopt_flags` to make sure `"number"` is already present in the user's setting. If it's not (e.g. `cursorlineopt=line`) then we need to disable `'cursorline'`.

Also if the user's value of `'cursorlineopt'` is already just "number" we can avoid the string allocation.

```suggestion
  if (curwin->w_p_culopt_flags & CULOPT_NBR) {
    if (strcmp(curwin->w_p_culopt, "number")) {
        save_w_p_culopt = curwin->w_p_culopt;
        curwin->w_p_culopt = (char_u *)xstrdup("number");
    }
    curwin->w_p_culopt_flags = CULOPT_NBR;
  } else {
    curwin->w_p_cul = false;
  }
```